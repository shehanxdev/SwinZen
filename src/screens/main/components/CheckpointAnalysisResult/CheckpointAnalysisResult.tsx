import React from 'react';
import { Pressable, View } from 'react-native';

import {
  FailCheckpointCrossIcon,
  MoveRightArrowIcon,
  PassCheckpointCrossIcon,
  PassStarCircleIcon,
  Text,
} from '@sz/components';
import { tw } from '@sz/config';
import { TextVariant } from '@sz/constants';

export interface CheckpointAnalysisResultProps {
  overallStatus: 'pass' | 'fail';
  checkPoint: string;
  subCheckpoint: string;
  onPress?: () => void;
}

export function CheckpointAnalysisResult({
  overallStatus,
  checkPoint,
  subCheckpoint,
  onPress,
}: CheckpointAnalysisResultProps) {
  const renderLeftOverallStatusIcon = () =>
    overallStatus === 'pass' ? <PassCheckpointCrossIcon /> : <FailCheckpointCrossIcon />;

  const renderRightActionIcon = () => (overallStatus === 'pass' ? <PassStarCircleIcon /> : <MoveRightArrowIcon />);

  return (
    <Pressable
      testID="CheckpointAnalysisResultTestID"
      style={tw`flex flex-row justify-between items-center h-8.5 w-full`}
      onPress={() => {
        if (overallStatus !== 'pass') {
          onPress();
        }
      }}>
      <View style={tw`flex flex-row items-center h-full`}>
        {renderLeftOverallStatusIcon()}
        <View style={tw`ml-9.25`}>
          <Text variant={TextVariant.Body2SemiBold}>{checkPoint}</Text>
        </View>
      </View>
      <View style={tw`flex flex-row items-center`}>
        {/* Width of this component and horizontal padding are not consistant in design system */}
        <View style={tw`justify-center rounded-2.5 mr-5 bg-Primary-Sz900 h-8.5 px-2.5`}>
          <Text variant={TextVariant.Labels}>{subCheckpoint}</Text>
        </View>
        {renderRightActionIcon()}
      </View>
    </Pressable>
  );
}
