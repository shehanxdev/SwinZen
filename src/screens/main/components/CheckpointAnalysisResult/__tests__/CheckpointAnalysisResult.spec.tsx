import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import { CheckpointAnalysisResult, CheckpointAnalysisResultProps } from './../CheckpointAnalysisResult';

describe('CheckpointAnalysisResult', () => {
  const testID = 'CheckpointAnalysisResultTestID';
  const onPressMock = jest.fn();

  const getRenderedCheckpointAnalysisResultComponent = (props?: Partial<CheckpointAnalysisResultProps>) =>
    render(
      <CheckpointAnalysisResult
        checkPoint="Checkpoint"
        subCheckpoint="Sub Checkpoint"
        onPress={onPressMock}
        overallStatus={'fail'}
        {...props}
      />,
    );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should find the component via testID', () => {
    const { getByTestId } = getRenderedCheckpointAnalysisResultComponent();
    const passwordField = getByTestId(testID);
    expect(passwordField).toBeTruthy();
  });

  test(`should render correctly`, () => {
    const rendered = getRenderedCheckpointAnalysisResultComponent();
    const renderedTree = rendered.toJSON();
    expect(renderedTree).toMatchSnapshot();
  });

  it('should trigger onPress callback when overallStatus is fail', () => {
    const { getByTestId } = getRenderedCheckpointAnalysisResultComponent({ overallStatus: 'fail' });

    const pressableComponent = getByTestId(testID);

    fireEvent.press(pressableComponent);

    expect(onPressMock).toHaveBeenCalled();
  });

  it('should not trigger onPress callback when overallStatus is pass', () => {
    const { getByTestId } = getRenderedCheckpointAnalysisResultComponent({ overallStatus: 'pass' });

    const pressableComponent = getByTestId(testID);

    fireEvent.press(pressableComponent);

    expect(onPressMock).not.toHaveBeenCalled();
  });
});
