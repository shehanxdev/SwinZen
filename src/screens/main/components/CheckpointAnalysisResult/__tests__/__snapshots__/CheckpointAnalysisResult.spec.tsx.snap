// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckpointAnalysisResult should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "height": 34,
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
  testID="CheckpointAnalysisResultTestID"
>
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "height": "100%",
      }
    }
  >
    <SvgMock
      fill="#F65815"
      height={34}
      width={34}
    />
    <View
      style={
        {
          "marginLeft": 37,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#F8F9FA",
              "lineHeight": 24,
            },
            {
              "fontFamily": "SourceSansPro-SemiBold",
              "fontSize": 16,
              "fontWeight": "600",
              "textAlign": "center",
            },
          ]
        }
      >
        Checkpoint
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#070707",
          "borderRadius": 10,
          "height": 34,
          "justifyContent": "center",
          "marginRight": 20,
          "paddingLeft": 10,
          "paddingRight": 10,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#F8F9FA",
              "lineHeight": 24,
            },
            {
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 15,
              "fontWeight": "600",
              "textAlign": "center",
            },
          ]
        }
      >
        Sub Checkpoint
      </Text>
    </View>
    <SvgMock
      fill="#FFFFFF"
      height={15}
      width={15}
    />
  </View>
</View>
`;
